File: lang/c/test/fun/vac/algorithms/meta/SearchTest.c
#include <fun/vac/algorithms/meta/Search.h>
#include <fun/vac/util/SequenceBuilder.h>
#include <fun/vac/util/TestRunner.h>

bool testBinarySearch(void) {
    size_t size = 32768;

    int arr[size];
    SequenceBuilder_packIncreasing(arr, size);

    if (Search_binarySearch(arr, size, -1) != size) {
        return false;
    }

    if (Search_binarySearch(arr, size, 2147483647) != size) {
        return false;
    }

    for (size_t i = 0; i < size; i++) {
        if (Search_binarySearch(arr, size, arr[i]) != i) {
            return false;
        }
    }

    return true;
}

bool testLinearSearch(void) {
    size_t size = 32768;

    int arr[size];
    SequenceBuilder_packIncreasing(arr, size);

    if (Search_linearSearch(arr, size, -1) != size) {
        return false;
    }

    if (Search_linearSearch(arr, size, 2147483647) != size) {
        return false;
    }

    for (size_t i = 0; i < size; i++) {
        if (Search_linearSearch(arr, size, arr[i]) != i) {
            return false;
        }
    }

    return true;
}

int main(void) {
    TestRunner_parseTest(testBinarySearch());

    TestRunner_parseTest(testLinearSearch());
}

File: go/muse/algorithms/meta/Search/Search_test.go
package Search

import (
    "testing"

    "muse/util/SequenceBuilder"
)

func TestBinarySearch(t *testing.T) {
    size := 32768

    arr := make([]int, size)
    SequenceBuilder.PackIncreasing(arr)

    if BinarySearch(arr, -1) != size {
        t.FailNow()
    }

    if BinarySearch(arr, 2147483647) != size {
        t.FailNow()
    }

    for i, v := range arr {
        if BinarySearch(arr, v) != i {
            t.FailNow()
        }
    }
}

func TestLinearSearch(t *testing.T) {
    size := 32768

    arr := make([]int, size)
    SequenceBuilder.PackIncreasing(arr)

    if LinearSearch(arr, -1) != size {
        t.FailNow()
    }

    if LinearSearch(arr, 2147483647) != size {
        t.FailNow()
    }

    for i, v := range arr {
        if LinearSearch(arr, v) != i {
            t.FailNow()
        }
    }
}
